Add PHP >= 8 support + updated tests and workflow.

josh-gaby opened the pull request:
Includes updated tests and workflow for running tests on PHP 5.4 -> 8.2

Also, due to symfony/http-foundation v6+ including return types and Ratchet supporting PHP < 7.0, there is a not very nice hack involving maintaining two versions of two classes (VirtualProxy and VirtualSessionStorage). This is currently working and passing all tests but definitly needs a better solution long-term.

josh-gaby committed:
Create test

josh-gaby committed:
Add missing properties to IoConnection.php

josh-gaby committed:
Update to work with PHP 8.2

josh-gaby committed:
Update tests to run on version upto 8.2

josh-gaby committed:
Merge pull request #1 from josh-gaby/php82

Php82

josh-gaby committed:
Allow manual workflow run

josh-gaby committed:
Merge pull request #2 from josh-gaby/php82

Allow manual workflow run

josh-gaby committed:
Update comments

josh-gaby committed:
Update tests

josh-gaby committed:
Merge pull request #3 from josh-gaby/php82

Update comments

josh-gaby committed:
Update PHPUnit to v9

josh-gaby committed:
Update PHPUnit to v9

josh-gaby committed:
Update workflow

josh-gaby committed:
Update workflow

josh-gaby committed:
Update workflow

josh-gaby committed:
Update workflow

josh-gaby committed:
Update workflow

josh-gaby committed:
Update workflow

josh-gaby committed:
Update workflow

josh-gaby committed:
Update workflow

josh-gaby committed:
Update workflow

josh-gaby committed:
Update workflow

josh-gaby committed:
Update workflow

josh-gaby committed:
Update workflow

josh-gaby committed:
Update workflow

josh-gaby committed:
Update workflow

josh-gaby committed:
Update workflow

josh-gaby committed:
Update workflow

josh-gaby committed:
Update workflow

josh-gaby committed:
Update workflow

josh-gaby committed:
Update workflow

josh-gaby committed:
Update workflow

josh-gaby committed:
Update workflow

josh-gaby committed:
Update workflow

josh-gaby committed:
Update workflow

josh-gaby committed:
Update workflow

josh-gaby committed:
Update test to use @before/@after annotations instead of setUp/tearDown

josh-gaby committed:
Remove invalid configurations and add missing @covers

josh-gaby committed:
Remove unused syntaxCheck from PHPUnit config

josh-gaby committed:
Remove usage of symfony/http-foundation >= 6

josh-gaby committed:
Test switching to createMock

josh-gaby committed:
Test switching to expectedException

josh-gaby committed:
Test switching to createMock

josh-gaby committed:
Revert to getMock

josh-gaby committed:
Revert to getMock

josh-gaby committed:
Update checkout version

josh-gaby committed:
Test extended testcase to allow switching between getMock/createMock

josh-gaby committed:
Test extended testcase to allow switching between getMock/createMock

josh-gaby committed:
Test extended testcase to allow switching between getMock/createMock

josh-gaby committed:
Test extended testcase to allow switching between getMock/createMock

josh-gaby committed:
Test extended testcase to allow switching between getMock/createMock

josh-gaby committed:
Add switches to support PHPUnit < 6 and > 6

josh-gaby committed:
Add switches to support PHPUnit < 6 and > 6

josh-gaby committed:
Add switches to support PHPUnit < 6 and > 6

josh-gaby committed:
Update to getMockBuilder

josh-gaby committed:
Revert some getMockBuilder

josh-gaby committed:
Update TopicTest

josh-gaby committed:
Update TopicTest

josh-gaby committed:
Update Connection

josh-gaby committed:
Revert changes

josh-gaby committed:
Revert changes

josh-gaby committed:
Update asserts

josh-gaby committed:
Try list() for array destructuring in testConnIsRemovedFromTopicOnClose

josh-gaby committed:
list() for array destructuring

josh-gaby committed:
Merge pull request #4 from josh-gaby/php82

Php82

josh-gaby committed:
Remove workflow_dispatch

josh-gaby committed:
Merge pull request #5 from josh-gaby/php82

Remove workflow_dispatch

josh-gaby committed:
Remove unnecessary include from HttpServer

josh-gaby committed:
Merge pull request #6 from josh-gaby/php82

Remove unnecessary include from HttpServer

josh-gaby committed:
Add http-foundation 6.0 back

josh-gaby committed:
Test return types

josh-gaby committed:
Allow manual work flow runs

josh-gaby committed:
Horrible hack to allow symfony/http-foundation >= 6.0 (definitely need a better solution)

josh-gaby committed:
Update to horrible hack to allow symfony/http-foundation >= 6.0 while maintaining PHP < 7.0 support.

josh-gaby committed:
Merge pull request #7 from josh-gaby/php82

Add http-foundation 6.0 back

josh-gaby committed:
Allow the WsConnection close function to include the optional reason.

josh-gaby committed:
Update ci.yml

Disable scheduled action run

josh-gaby committed:
Update ci.yml

Disable scheduled action run

CViniciusSDias commented:
Hey there, @cboden 
This PR could re-live the project.
:-D

neoteknic commented:
why php5.4 its useless... just drop php < 8

jaxwilko commented:
@josh-gaby awesome work, hopefully it can get reviewed and merged soon! :)

josh-gaby committed:
Add missing __isset function to the IoConnection class.

Tofandel commented on src/Ratchet/Http/HttpRequestParser.php:
```diff
@@ -19,6 +19,40 @@ class HttpRequestParser implements MessageInterface {
+            return $this->_properties[$key];
+        }
+    }
+
```
What about `__isset`  ?

josh-gaby replied:
I'll add a trait for supporting dynamic properties and will include all the get, set, and isset functions

Tofandel commented on src/Ratchet/Server/IoServer.php:
```diff
@@ -80,61 +114,61 @@ public function run() {
     public function handleConnect($conn) {
-        $conn->decor = new IoConnection($conn);
-        $conn->decor->resourceId = (int)$conn->stream;
+        $io_conn = new IoConnection($conn);
```
I'm guessing `$conn->decor` should still be left to avoid a breaking change

josh-gaby replied:
decor is not a property of \React\Socket\ConnectionInterface so can't be left in because php 8 deprecated dynamic property creation.

Tofandel replied:
Then maybe we also make a PR to reactphp/socket to add #[AllowDynamicProperties] to the interface https://github.com/reactphp/socket/blob/3.x/src/ConnectorInterface.php

Tofandel replied:
Or we simply use spl storage locally in IoServer to store the decor and retrieve it when needed

Tofandel commented on src/Ratchet/Server/IoServer.php:
```diff
@@ -80,61 +114,61 @@ public function run() {
+     * @param \Ratchet\ConnectionInterface $io_conn
      */
-    public function handleData($data, $conn) {
+    public function handleData($data, $io_conn) {
```
Changing the parameter there is a breaking change, because IoServer can be extended

josh-gaby replied:
I agree, this a breaking change, however it was done because \React\Socket\ConnectionInterface doesn't actually have the decor property previously being used to pass the connection around and PHP 8 deprecated dymanic property creation.

Tofandel commented on src/Ratchet/Session/Storage/Proxy/VirtualProxy.php:
```diff
@@ -1,54 +1,16 @@
+
+if (version_compare(\Composer\InstalledVersions::getVersion('symfony/http-foundation'), '6.0.0') === -1) {
+    // The version of http-foundation is < 6, include a class without return types
+    include 'VirtualProxy-HttpFoundation<6.php';
```
This is unnecessary, the return types can be added without breaking the signature, it's removing or changing them that's an issue

josh-gaby replied:
This is required if support for PHP < 7.0 is to be maintained, earlier versions didn't support return types.

Tofandel replied:
Why keep support for such old php? that's asking for maintenance hell, they have been EOL for more than 6 years

josh-gaby replied:
I agree, but like I told the commenter above, it's not my project, so all I'm doing is getting it working again for php 8+

Personally, I would do it as a new major version that supported php >= 7.4 and leave < 7.4 using the current version but that's nothing to do with me.

Also, this pull request has been sitting here for a year now, I'm not sure how likely it is to even get merged at this point to be honest.

Tofandel replied:
There is ongoing work recently on ratchet from the maintainers to get the projects back alive

See

https://github.com/ratchetphp/Ratchet/issues/1054

So it's quite likely

It's okay if it's not your project, you can do a PR with a breaking change and mark it as such so it will be released as a major, otherwise those kind of big PR's are a maintenance pain and it's actually more likely to be accepted if you don't add breaking changes in your PR but drop support for older php versions

But since you made a compat layer already, you can leave it in, to be removed on the next major

Tofandel commented on src/Ratchet/WebSocket/WsServer.php:
```diff
@@ -61,6 +61,13 @@ class WsServer implements HttpServerInterface {
+     *
+     * @var array
+     */
+    protected $_properties = [];
```
Seeing this is reused in a lot of places, this should be a trait or `AllowDynamicProperties` can be used instead

josh-gaby replied:
`#[AllowDynamicProperties]` is only available after 8.0 but traits may be a better option

Tofandel replied:
That's not an issue, php attributes are made to be backwards compatible all the way to php 4, because they are treated as comments otherwise and an attribute class that doesn't exist is simply ignored, this was thought specifically so that DynamicProperties support could be kept by only adding this attribute without being a breaking change

josh-gaby replied:
I didn't think about that, I've already switched it to a trait.

josh-gaby committed:
refactor: switch to using a dynamic properties trait

josh-gaby committed:
refactor: switch IoConnection to use the new dynamic properties trait

josh-gaby committed:
ci: re-enable running ci on a schedule

josh-gaby committed:
Merge branch 'master' into php82

josh-gaby committed:
Merge pull request #8 from josh-gaby/php82

Php82

NigelCunningham committed:
Provide a way to set and unset a dynamic property.

(Previously, the set test checked for existence, giving you no way
of creating the property dynamically).

Signed-off-by: Nigel Cunningham <nigel@nigelcunningham.com.au>

dvdknaap commented:
what is still needed to merge this PR ? @josh-gaby  @clue ?

josh-gaby committed:
Merge pull request #9 from NigelCunningham/master

Provide a way to set and unset a dynamic property.

dvdknaap commented:
@josh-gaby i don't get why you still havent received any reply yet from the maintainers. i see your still updating your fork. 
i'm gonna test it in my project. will let you know if i find something

josh-gaby commented:
@dvdknaap I've been using it in a reasonably sized project for the last year, haven't had any issues at all so far but if you do find anything let me know and I'll try fix it

dvdknaap commented:
> @dvdknaap I've been using it in a reasonably sized project for the last year, haven't had any issues at all so far but if you do find anything let me know and I'll try fix it

I also tested it and it;s working without issues !

Could you maybe add support for SF 7 so i can finally upgrade ? :D

josh-gaby committed:
feat: add support for symfony v7

josh-gaby committed:
Merge pull request #10 from josh-gaby/php82

feat: add support for symfony v7

josh-gaby committed:
Merge branch 'ratchetphp:master' into master

dvdknaap commented:
@josh-gaby thanks for the update. It's working without any issues ! 🥳😎

zhiyong-ft commented:
@dvdknaap could you please elaborate on how to integrate @josh-gaby's fork into a project? I tried something like the following:
```
    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/josh-gaby/Ratchet.git"
        }
    ],
```
It doesn't work.

josh-gaby commented:
@zhiyong-ft 
I'm loading using the 'git' type:

```
"repositories": [
    {
      "url": "https://github.com/josh-gaby/Ratchet.git",
      "type": "git"
    }
  ],
  ```
  
  and then specifying the branch in the requirements like this `"cboden/ratchet": "dev-php82"`

zhiyong-ft commented:
> "cboden/ratchet": "dev-php82"

This is what I missed. Thanks.

dvdknaap commented:
> 
> 
> 
> > "cboden/ratchet": "dev-php82"
> 
> This is what I missed. Thanks.

I just use dev-master as I can see that branch is already merged so it's safe to move to the master branch.

Tofandel commented:
Tip you can use

"cboden/ratchet": "dev-master as 0.5.0"

If you have a version conflict somewhere

sederther commented:
[src/Ratchet/Session/Storage/VirtualSessionStorage-HttpFoundation<6.php](https://github.com/ratchetphp/Ratchet/pull/1014/files#diff-f22acebd417d9f44c4f483c0bc795ebab43c5647dad01984d0ff59bcd72bb7de)

"less then" (<) sign must not be used since the character is forbidden in file names on Windows OS.

Tofandel commented:
And gently reminder that php 5.6 has been EOL for 6+ years and 7.4 for 2+ years, so there is no reason to keep a compatibility layer, support should be dropped

https://www.php.net/eol.php

PaulRotmann commented:
Hey! Thanks for this comprehensive PR and all the great ideas you've shared! 🙏

We're giving Ratchet some much-needed love and attention! As part of our major Ratchet revival initiative (#1054), we've been working through many of the improvements and ideas you've proposed here. Several of the topics you've addressed have already been incorporated into our ongoing modernization efforts.

Since you've closed this PR yourself and we're making significant progress on the broader revival project, we wanted to acknowledge the valuable contributions and thoughts you've shared. Many of your insights are helping guide our approach as we work through the comprehensive updates.

If you're interested in continuing to contribute or discuss any specific aspects that haven't been addressed yet, we'd love to continue the conversation in our GitHub [Discussions](https://github.com/ratchetphp/Ratchet/discussions). Just reference this PR when you post.

If you want to support Ratchet's revival, please consider [sponsoring](https://github.com/sponsors/reactphp) our work! ❤️
Thanks again for your thoughtful contribution to making Ratchet better!
